src/ert/ensemble_evaluator/evaluator_async.py

```diff
@@ -0,0 +1,474 @@
+
+from ert.serialization import evaluator_marshaller, evaluator_unmarshaller
+
+from ._builder import Ensemble
+from .config import EvaluatorServerConfig
+from .identifiers import (
+    EVGROUP_FM_ALL,
+    EVTYPE_EE_SNAPSHOT,
+    EVTYPE_EE_SNAPSHOT_UPDATE,
+    EVTYPE_EE_TERMINATED,
+    EVTYPE_EE_USER_CANCEL,
+    EVTYPE_EE_USER_DONE,
+    EVTYPE_ENSEMBLE_CANCELLED,
+    EVTYPE_ENSEMBLE_FAILED,
+    EVTYPE_ENSEMBLE_STARTED,
+    EVTYPE_ENSEMBLE_STOPPED,
+)
+from .snapshot import PartialSnapshot
+from .state import (
+    ENSEMBLE_STATE_CANCELLED,
+    ENSEMBLE_STATE_FAILED,
+    ENSEMBLE_STATE_STOPPED,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class EnsembleEvaluatorAsync:
+    def __init__(self, ensemble: Ensemble, config: EvaluatorServerConfig, iter_: int):
+        # Without information on the iteration, the events emitted from the
+        # evaluator are ambiguous. In the future, an experiment authority* will
+        # "own" the evaluators and can add iteration information to events they
+        # emit. In the meantime, it is added here.
+        # * https://github.com/equinor/ert/issues/1250
+        self._iter: int = iter_
+        self._config: EvaluatorServerConfig = config
+        self._ensemble: Ensemble = ensemble
+
+        self._loop: Optional[asyncio.AbstractEventLoop] = None
+
+        self._clients: Set[WebSocketServerProtocol] = set()
+        self._dispatchers_connected: asyncio.Queue[None] = asyncio.Queue()
+
+        self._events: asyncio.Queue[CloudEvent] = asyncio.Queue()
+        self._messages_to_send: asyncio.Queue[str] = asyncio.Queue()
+
+        self._result = None
+
+        self._ee_tasks: List[asyncio.Task[None]] = []
+        self._server_started: asyncio.Event = asyncio.Event()
+        self._server_done: asyncio.Event = asyncio.Event()
+
+        # batching section
+        self._batch_processing_queue: asyncio.Queue[
+            List[Tuple[Callable[[List[CloudEvent]], Awaitable[None]], CloudEvent]]
+        ] = asyncio.Queue()
+        self._max_batch_size: int = 500
+        self._batching_interval: int = 2
+
+    async def _publisher(self) -> None:
+        while True:
+            msg = await self._messages_to_send.get()
+            await asyncio.gather(
+                *[client.send(msg) for client in self._clients],
+                return_exceptions=True,
+            )
+            self._messages_to_send.task_done()
+
+    async def _append_message(self, snapshot_update_event: PartialSnapshot) -> None:
+        message = self._create_cloud_message(
+            EVTYPE_EE_SNAPSHOT_UPDATE,
+            snapshot_update_event.to_dict(),
+        )
+        if message:
+            await self._messages_to_send.put(message)
+
+    async def _process_event_buffer(self) -> None:
+        while True:
+            batch = await self._batch_processing_queue.get()
+            function_to_events_map: Dict[
+                Callable[[List[CloudEvent]], Awaitable[None]], List[CloudEvent]
```

Comment:
The inner type is used here and on line 82. Can we put it outside the class for reuse and give it a more descriptive name?

Reply:
What to put it outside the class?

Reply:
We can use it as a type alias, that is how we did with the long composite types in the new drivers. We could also have it as a class variable.